README.pod
=pod

=encoding utf-8

=head1 NAME

Metabase - A database framework and API for resource metadata

=head1 VERSION

version 1.002

=head1 DESCRIPTION

Metabase is a database framework and API for resource metadata.  The framework
describes how arbitrary data ("facts") are associated with particular resources
or related to each other.  The API describes how to store, retrieve, and search
this information.

=head2 History and Motivation

Metabase was originally designed as a means of storing reports from the CPAN
Testers project.  When Metabase was initially developed, CPAN Testers reports
were sent by individual testers to a single email server, which then forwarded
them to a USENET group, which was considered the authoritative store.  This
presented problems: some testers couldn't send email, the system wasn't very
searchable, was hard to mirror, and the data inside the system was entirely
unstructured.

Metabase aimed to avoid all of those problems by being transport-neutral,
searchable and easier to mirror by design, and geared toward storing structured
data.  Simplicity is another design goal: while it has several moving parts,
they're all simple and designed to be replaceable and extensible, rather than
to be a perfect design up front.

=head1 OVERVIEW

A Metabase has several parts:

=over

=item *

L<Metabase::Librarian>, a class which manages access to the Archive and Index

=item *

L<Metabase::Gateway>, a role for managing submission of facts to the Librarian

=item *

L<Metabase::Archive>, a role for storing and retrieving facts

=item *

L<Metabase::Index>, a role for indexing and searching facts

=item *

L<Metabase::Query>, a role for translating a common query data structure into a
backend-specific query

=back

Roles require implementations.  These could use flat files, relational
databases, object databases, cloud services, or anything else that can satisfy
the role API.

Metabase comes with some simple, stupid backends for testing:

=over 4

=item *

L<Metabase::Archive::Filesystem>

=item *

L<Metabase::Index::FlatFile>

=back

Facts stored with in a Metabase are defined as subclasses of L<Metabase::Fact>.
L<Metabase::Report> is a subclass that relates multiple facts.

L<Metabase::Web> provides the web API for storing, searching and retrieving
facts.  L<Metabase::Client::Simple> is the client library to submit facts to a
Metabase::Web server.  A future Metabase::Client class will provide submit and
search capabilities.

=head1 SEE ALSO

=over 4

=item *

L<Metabase::Backend::AWS>

=item *

L<Metabase::Backend::MongoDB>

=item *

L<Metabase::Backend::SQL>

=back

=for :stopwords cpan testmatrix url annocpan anno bugtracker rt cpants kwalitee diff irc mailto metadata placeholders metacpan

=head1 SUPPORT

=head2 Bugs / Feature Requests

Please report any bugs or feature requests through the issue tracker
at L<https://rt.cpan.org/Public/Dist/Display.html?Name=Metabase>.
You will be notified automatically of any progress on your issue.

=head2 Source Code

This is open source software.  The code repository is available for
public review and contribution under the terms of the license.

L<https://github.com/dagolden/metabase>

  git clone git://github.com/dagolden/metabase.git

=head1 AUTHORS

=over 4

=item *

David Golden <dagolden@cpan.org>

=item *

Ricardo Signes <rjbs@cpan.org>

=item *

Leon Brocard <acme@cpan.org>

=back

=head1 CONTRIBUTORS

=over 4

=item *

Florian Ragwitz <rafl@debian.org>

=item *

Graham Knop <haarg@haarg.org>

=item *

Leon Brocard <acme@astray.com>

=item *

Ricardo SIGNES <rjbs@cpan.org>

=back

=head1 COPYRIGHT AND LICENSE

This software is Copyright (c) 2013 by David Golden.

This is free software, licensed under:

  The Apache License, Version 2.0, January 2004



Changes
Revision history for Metabase

{{$NEXT}}

1.003     2013-11-28 07:46:50 America/New_York
    - Switch from Class::MOP::load_class to Class::Load::load_class

1.002     2013-07-16 10:36:43 America/New_York
    - Fix a test that had been broken by 5.18's hash ordering changes

1.001     2012-05-18 15:34:47 America/New_York

    [FIXED]

    - Fixed bug where default CHI wasn't initialized properly

1.000     2012-05-16 13:23:04 America/New_York

    - The 1.000 release reflects a major reorganization of the Metabase
      distribution and is not backwards compatible.

    - The Archive and Index roles have an expanded API with additional
      required methods

    - A Query role has been added to abstract query requirements from the
      indexer to allow for backend-specific query implementations

    - Tests have been converted to Test::Routine style roles, which are
      released with the distribution for use in backend tests in other
      distributions.

    - AWS and SQLite backends have been removed and will be released as
      separate distributions.

    - Solr backend has been removed.

0.016     2010-12-07 15:27:49 EST5EDT

    - No changes.  Released as stable

0.015     2010-09-28 09:41:27 EST5EDT

    - Add explicit dependency on SQL::Translator 0.11006 for DBIx::Class
      deploy() used in Metabase::Archive::SQLite

0.014     2010-09-20 22:22:45 EST5EDT

    - Now uses SimpleDB::Client instead of SimpleDB::Class

    - Updated dist.ini to @DAGOLDEN style

    - Prerequisites now auto-detected

0.013     2010-09-01 18:29:24 EST5EDT

    - Improve diagnostics on failure

    - Fix occasional failures in t/librarian/report.t due to time
      comparison between original and saved reports

    - Add iterator generators to librarian (only supported on S3 so far)

    - Load fact classes on demand

    - Require latest Metabase::Fact (0.018)

0.012     2010-04-24 15:18:33 EST5EDT

    - Fix authentication cache bug (first check was always failing)

0.011     2010-04-23 23:38:58 EST5EDT

    - Fix cache attribute type constraint

    - Cache authentication strings, not objects

0.010     2010-04-23 17:40:39 EST5EDT

    - Add authentication caching.  Adds CHI as a prereq

    - Properly document Metabase::Gateway as a Moose::Role

0.009     2010-04-22 07:26:46 EST5EDT

    - Don't add undefined values to the index

0.008     2010-04-20 09:54:07 EST5EDT

    - Fact content metadata keys no longer have to be lowercase

0.007     2010-04-11 12:00:49 EST5EDT

    - Fix a bug in Metabase::Archive::S3 -- NOTE: THIS WILL BREAK
      ACCESS TO OLD S3 FACTS

0.006     2010-03-29 14:29:25 EST5EDT

    - JSON will now always encode/decode to/from ASCII

    - Converted to Dist::Zilla for easier release management

0.005

    - Remove dependency on JSON::XS and rely on JSON to use it
      if installed (see Task::Metabase)

    - Ensure JSON encoding includes utf8 encoding to avoid
      errors compressing wide characters in Metabase::Archive::S3

0.004 Mon Mar 15 22:30:34 EDT 2010
    - Fixed -limit operator for SimpleDB search
    - Reduced numerous prerequisites to lower verisons

0.003 Mon Mar  8 23:42:52 EST 2010
    - heavily revised for CPAN Testers 2.0 effort
    - Metabase::Gateway is now a Moose::Role
    - Added Amazon Web Service backends

0.002 (never released)
    - private development version, never released to CPAN

0.001 24 Jun 2009
    - original skeleton created at Oslo QA hackathon 2008


lib/Metabase/Gateway.pm
use 5.006;
use strict;
use warnings;

package Metabase::Gateway;
# ABSTRACT: Manage Metabase fact submission
# VERSION

use Moose::Role;

use Metabase::Fact 0.018;
use Metabase::Librarian;
use Metabase::User::Profile;
use Metabase::User::Secret;
use CHI;
use namespace::autoclean;

requires '_build_public_librarian';
requires '_build_private_librarian';
requires '_build_fact_classes';

has public_librarian => (
  is       => 'ro',
  isa      => 'Metabase::Librarian',
  lazy     => 1,
  builder  => '_build_public_librarian',
);

has private_librarian => (
  is       => 'ro',
  isa      => 'Metabase::Librarian',
  lazy     => 1,
  builder  => '_build_private_librarian',
);

# XXX life becomes a lot easier if we say that fact classes MUST have 1-to-1
# relationship with a .pm file. -- dagolden, 2009-03-31

has fact_classes => (
  is  => 'ro',
  isa => 'ArrayRef[Str]',
  auto_deref => 1,
  lazy     => 1,
  builder => '_build_fact_classes',
);

has approved_types => (
  is          =>  'ro',
  isa         =>  'ArrayRef[Str]',
  auto_deref  => 1,
  lazy        => 1,
  builder     => '_build_approved_types',
  init_arg    => undef,
);

has disable_security => (
  is          => 'ro',
  isa         => 'Bool',
  default     => 0,
);

has allow_registration => (
  is          => 'ro',
  isa         => 'Bool',
  default     => 1,
);

has authentication_timeout => (
  is          => 'ro',
  isa         => 'Int',
  default     => 600,
);

has cache_options => (
  is          => 'ro',
  isa         => 'HashRef',
  default     => sub { my $hash = {}; return { driver => 'Memory', datastore => $hash } },
);

has _cache => (
  is          => 'ro',
  isa         => 'CHI::Driver',
  lazy        => 1,
  builder     => '_build_cache',
);

# create a CHI cache
sub _build_cache {
  my ($self) = @_;
  return CHI->new( %{ $self->cache_options } );
}

# recurse report classes -- less to specify to new()
sub _build_approved_types {
  my ($self) = @_;
  my @queue = $self->fact_classes;
  my @approved;
  while ( my $class = shift @queue ) {
    push @approved, $class;
    # XXX $class->can('fact_classes') ?? -- dagolden, 2009-03-31
    push @queue, $class->fact_classes if $class->isa('Metabase::Report');
  }
  return [ map { Class::Load::load_class($_); $_->type } @approved ];
}

# for use in handle_XXXX methods
sub _fatal {
  my ($self, $code, $reason, $details) = @_;
  $code ||= 500;
  $reason ||= "unknown error";
  $details ||= '';
  chomp $details;
  my $message = "$code\: $reason";
  $message .= ": $details" if $details;
  die "$message\n";
}

sub _validate_resource {
  my ($self, $request) = @_;

  # XXX Well... yeah, eventually we'll want to reject reports for dists that
  # don't, you know, exist. -- rjbs, 2008-04-06
  1;
}

sub _validate_submitter {
  my ($self, $user_guid, $user_secret) = @_;

  return 1 if $self->disable_security;

  # did we get arguments?
  die "no user identity provided\n"
    unless $user_guid;
  die "no user secret provided\n"
    unless $user_secret;

  my $user_resource = "metabase:user:$user_guid";

  my $secret = $self->_cache->get("secret/$user_guid");

  if ( ! defined $secret ) {
    # check if we have a secret on file
    eval {
      my $found = $self->private_librarian->search(
          'core.type' => 'Metabase-User-Secret',
          'core.resource' => $user_resource,
      );
      unless ( defined $found->[0] ) {
        die "credentials for $user_resource not found\n";
      }
      my $obj = $self->private_librarian->extract($found->[0]);
      # if we haven't died, we have it, so cache it
      $secret = $obj->content;
      $self->_cache->set(
        "secret/$user_guid", $secret, $self->authentication_timeout
      );
    };
  }

  # match against submitted secret
  die "authentication failed for $user_resource\n"
    unless defined $secret && $user_secret eq $secret;

  # submitter is good!
  return 1;
}

sub _validate_fact_struct {
  my ($self, $struct, @approved) = @_;

  # exists and has type
  die "no fact provide" unless defined $struct;
  die "fact type not provided" unless defined $struct->{metadata}{core}{type};

  # approved fact type
  my $type = $struct->{metadata}{core}{type};
  unless ( grep { $type eq $_ } @approved ) {
    die "$type is not an approved fact type\n";
  }

  # has content
  die "no content provided" unless defined $struct->{content};

  # required metadata
  for my $key ( qw/resource type schema_version guid creator creation_time/ ) {
    my $meta = $struct->{metadata}{core}{$key};
    die "no '$key' provided in core metadata"
      unless defined $meta;
    # XXX really should check meta validity: [ //str => 'abc' ], but lets wait
    # until we decide on sugar for metadata types -- dagolden, 2009-03-31
  }

  die "submissions must not include resource or content metadata"
    if $struct->{metadata}{content} or $struct->{metadata}{resource};

  return 1;
}

sub _check_permissions {
  my ($self, $user_guid, $action, $fact) = @_;

  # The devil may care, but we don't. -- rjbs, 2009-03-30

  # E.g. do we let a user submit a fact they aren't listed as the creator for?
  # -- dagolden, 2010-02-28

  return 1;
}

sub _thaw_fact {
  my ($self, $struct, @approved) = @_;

  $self->_validate_fact_struct($struct, @approved);
  my $type = $struct->{metadata}{core}{type};
  return Metabase::Fact->class_from_type($type)->from_struct($struct);
}

# NOTE ON ERRORS: die with _fatal( XXX => reason => details )
sub handle_submission {
  my ($self, $struct, $user_guid, $user_secret) = @_;

  # use Data::Dumper;
  # local $SIG{__WARN__} = sub { warn "@_: " . Dumper($struct); };

  # authenticate
  unless ( eval { $self->_validate_submitter( $user_guid, $user_secret ); 1 } ) {
    $self->_fatal( 401 => "unauthorized" => $@ );
  }

  # thaw
  my $fact = eval { $self->_thaw_fact($struct, $self->approved_types) };
  unless ( $fact ) {
    $self->_fatal( 400 => "invalid submission data" => $@ );
  }

  # action allowed for this submitter for this fact
  unless ( eval { $self->_check_permissions($user_guid => submit => $fact) } ) {
    $self->_fatal( 400 => "cannot accept fact from current submitter" => $@ );
  }

  # accepted by librarian
  my $guid = eval { $self->enqueue($fact) };
  unless ( $guid ) {
    $self->_fatal( 500 => "submission was not stored" => $@ );
  }

  return $guid;
}

# NOTE ON ERRORS: die with _fatal( XXX => reason => details )
sub handle_registration {
  my ($self, $profile_struct, $secret_struct) = @_;

  $self->_fatal( 400 => "new user registration disabled" )
    unless $self->allow_registration;

  # thaw profile
  my $profile = eval { $self->_thaw_fact($profile_struct, 'Metabase-User-Profile') };
  unless ( $profile ) {
    $self->_fatal( 400 => "invalid registration profile" => $@ );
  }

  # thaw secret
  my $secret = eval { $self->_thaw_fact($secret_struct, 'Metabase-User-Secret') };
  unless ( $secret ) {
    $self->_fatal( 400 => "invalid registration credential" => $@ );
  }

  # neither should exist
  if ( $self->public_librarian->exists( $profile->guid )
    || $self->private_librarian->exists( $secret->guid )
  ) {
    $self->_fatal( 400 => "already registered" );
  }

  # store with respective librarians
  my ($secret_guid, $profile_guid);
  eval {
    $secret_guid = $self->private_librarian->store( $secret );
    $profile_guid = $self->public_librarian->store( $profile );
  };
  unless ( $secret_guid && $profile_guid ) {
    $self->_fatal( 500 => "registration failed" => $@ );
  }

  # profile accepted by librarian
  return $profile_guid;
}

sub enqueue {
  my ($self, $fact) = @_;
  return $self->public_librarian->store($fact);
}


1;

__END__

=head1 SYNOPSIS

=head2 Creating a Gateway class

  use strict;
  use warnings;
  package Custom::Metabase;

  use Moose;
  with 'Metabase::Gateway';

  sub _build_fact_classes { ... }
  sub _build_public_librarian { ... }
  sub _build_private_librarian { ... }

=head2 Using a Gateway class

  my $mg = Custom::Gateway->new( @args );
  $mg->handle_submission( $fact_struct, $user_guid, $user_secret);

=head1 DESCRIPTION

The Metabase::Gateway role manages submissions to the Metabase.  It
provides fact and submitter validation or authorization before storing
new facts in a Metabase.

=head1 USAGE

=head2 Required methods

The following builders must be provided by the class that applies this role.
All three are lazy.

=head3  C<_build_fact_classes>

This builder must return an array reference of Metabase::Fact subclasses.

=head3  C<_build_public_librarian>

=head3  C<_build_private_librarian>

These builders must return Metabase::Librarian objects.

=head2 Configurable attributes

=head3 C<disable_security>

A boolean option.  If true, submitter profiles will not be authenticated.
(This is generally useful for testing, only.) Default is false.

=head3 C<allow_registration>

A boolean option.  If true, new submitter profiles and secrets may be
stored. Default is true.

=head3 C<authentication_timeout>

Number of seconds to cache user authentication data.  Defaults to 600.

=head3 C<cache_options>

A hash reference of constructor arguments for a L<CHI> cache. Defaults to
C<< { driver => 'Memory' } >>.

=head2 Generated attributes

=head3 C<approved_types>

Returns a list of approved fact types.  Used for validating submitted facts.

A "type" is a class name with "::" converted to "-", so this attribute
returns an arrayref of the C<fact_classes> attribute converted to types.

=head3 C<fact_classes>

Array reference containing a list of valid L<Metabase::Fact> subclasses. Only
facts from these classes may be added to the Metabase.

=head3 C<public_librarian>

A librarian object to manage fact data.

=head3 C<private_librarian>

A librarian object to manage user authentication data and possibly other
facts that should be segregated from searchable and retrievable facts.
This should not be the same as the public_librarian.

=head2 Methods provided

=head3 C<enqueue>

  $mg->enqueue( $fact );

Add a fact from a user (identified by a profile) via the public_librarian.
Used internally by handle_submission.

=head3 C<handle_submission>

  $mg->handle_submission( $fact_struct, $user_guid, $user_secret);

Extract a fact a deserialized data structure and add it to the Metabase via the
public_librarian. The fact is regenerated from the C<as_struct> method.

=head3 C<handle_registration>

  $mg->handle_registration( $profile_struct, $secret_struct );

Extract a new user profile and secret from deserialized data structures
and add them via the public_librarian and private_librarian, respectively.

=cut

